packages/java/parser-jvm-plugin-backbone/src/main/java/com/vaadin/hilla/parser/plugins/backbone/JsonValuePlugin.java

```diff
@@ -0,0 +1,106 @@
+package com.vaadin.hilla.parser.plugins.backbone;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.vaadin.hilla.parser.core.AbstractPlugin;
+import com.vaadin.hilla.parser.core.Node;
+import com.vaadin.hilla.parser.core.NodeDependencies;
+import com.vaadin.hilla.parser.core.NodePath;
+import com.vaadin.hilla.parser.models.*;
+import com.vaadin.hilla.parser.plugins.backbone.nodes.TypeSignatureNode;
+import jakarta.annotation.Nonnull;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Adds support for Jackson's {@code JsonValue} and {@code JsonCreator}
+ * annotations.
+ */
+public class JsonValuePlugin
+        extends AbstractPlugin<BackbonePluginConfiguration> {
+    private final Map<Class<?>, Optional<Class<?>>> jsonValues = new HashMap<>();
+
+    @Override
+    public void enter(NodePath<?> nodePath) {
+    }
+
+    @Override
+    public void exit(NodePath<?> nodePath) {
+    }
+
+    @Nonnull
+    @Override
+    public NodeDependencies scan(@Nonnull NodeDependencies nodeDependencies) {
+        return nodeDependencies;
+    }
+
+    @Nonnull
+    @Override
+    public Node<?, ?> resolve(@Nonnull Node<?, ?> node,
+            @Nonnull NodePath<?> parentPath) {
+        if (node instanceof TypeSignatureNode typeSignatureNode) {
+            if (typeSignatureNode
+                    .getSource() instanceof ClassRefSignatureModel classRefSignatureModel) {
+                var cls = (Class<?>) classRefSignatureModel.getClassInfo()
+                        .get();
+                // Check if the class has the annotations which qualify for a
+                // value type. If so, replace the type with the corresponding
+                // value type.
+                Optional<TypeSignatureNode> valueNode = getValueType(cls)
+                        .map(SignatureModel::of).map(TypeSignatureNode::of);
+                return valueNode.orElse(typeSignatureNode);
+            }
+        }
+
+        return node;
+    }
+
+    private Optional<Class<?>> getValueType(Class<?> cls) {
+        // Use cached results to avoid recomputing the value type.
+        return jsonValues.computeIfAbsent(cls, this::findValueType);
+    }
+
+    private Optional<Class<?>> findValueType(Class<?> cls) {
+        // First of all, we check that the `@JsonValue` annotation is
+        // used on a method of the class.
+        var jsonValue = Arrays.stream(cls.getMethods())
+                .filter(method -> Arrays.stream(method.getAnnotations())
+                        .map(ann -> ann.annotationType().getName()).anyMatch(
+                                name -> name.equals(JsonValue.class.getName())))
```

Comment:
Couldn't the `method.isAnnotationPresent(JsonValue.class)` be used here instead of streaming the method annotations and checking the name equality?

Reply:
Good catch! Damn copy-paste.

Just for information, I haven't added the opposite test (only `JsonCreator`), since such an endpoint crashes the generator anyway.

Reply:
Definitely. This is a leftover of an initial different implementation. Thanks for the suggestion.